File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/conformance/consent_runner.rb
# frozen_string_literal: true

require "json"
require_relative "result"
require_relative "../consent"

module IAIso
  module Conformance
    module ConsentRunner
      module_function

      def run(spec_root)
        path = File.join(spec_root, "consent", "vectors.json")
        doc = ::JSON.parse(File.read(path))
        shared_key = doc["hs256_key_shared"]
        out = []
        (doc["scope_match"] || []).each { |v| out << run_scope_match(v) }
        (doc["scope_match_errors"] || []).each { |v| out << run_scope_match_error(v) }
        (doc["valid_tokens"] || []).each { |v| out << run_valid_token(shared_key, v) }
        (doc["invalid_tokens"] || []).each { |v| out << run_invalid_token(shared_key, v) }
        (doc["issue_and_verify_roundtrip"] || []).each { |v| out << run_roundtrip(shared_key, v) }
        out
      end

      def parse_alg(v)
        v["algorithm"] || IAIso::Consent::Algorithm::HS256
      end

      def run_scope_match(v)
        name = "scope_match/#{v["name"]}"
        got = IAIso::Consent::Scopes.granted(v["granted"], v["requested"])
        want = v["expected"]
        return VectorResult.fail("consent", name, "got #{got}, want #{want}") if got != want
        VectorResult.pass("consent", name)
      rescue => e
        VectorResult.fail("consent", name, "unexpected exception: #{e.message}")
      end

      def run_scope_match_error(v)
        name = "scope_match_errors/#{v["name"]}"
        expect = v["expect_error"]
        IAIso::Consent::Scopes.granted(v["granted"], v["requested"])
        VectorResult.fail("consent", name, "expected error containing '#{expect}', got Ok")
      rescue ArgumentError => e
        unless e.message.downcase.include?(expect.downcase)
          return VectorResult.fail("consent", name, "expected '#{expect}', got: #{e.message}")
        end
        VectorResult.pass("consent", name)
      end

      def run_valid_token(shared_key, v)
        name = "valid_tokens/#{v["name"]}"
        now = v["now"].to_i
        issuer = v["issuer"] || "iaiso"
        alg = parse_alg(v)
        verifier = IAIso::Consent::Verifier.new(
          algorithm: alg, issuer: issuer, hs_key: shared_key,
          clock: ->{ now },
        )
        scope = verifier.verify(v["token"])
        exp = v["expected"]
        return VectorResult.fail("consent", name, "sub: got #{scope.subject}, want #{exp["sub"]}") if scope.subject != exp["sub"]
        return VectorResult.fail("consent", name, "jti: got #{scope.jti}, want #{exp["jti"]}") if scope.jti != exp["jti"]
        if scope.scopes != exp["scopes"]
          return VectorResult.fail("consent", name, "scopes mismatch: got #{scope.scopes}, want #{exp["scopes"]}")
        end
        want_exec = exp["execution_id"]
        if scope.execution_id != want_exec
          return VectorResult.fail("consent", name, "execution_id: got #{scope.execution_id.inspect}, want #{want_exec.inspect}")
        end
        VectorResult.pass("consent", name)
      rescue => e
        VectorResult.fail("consent", name, "verify failed: #{e.message}")
      end

      def run_invalid_token(shared_key, v)
        name = "invalid_tokens/#{v["name"]}"
        now = v["now"].to_i
        issuer = v["issuer"] || "iaiso"
        alg = parse_alg(v)
        exec_id = v["execution_id"]
        expect = v["expect_error"]
        verifier = IAIso::Consent::Verifier.new(
          algorithm: alg, issuer: issuer, hs_key: shared_key,
          clock: ->{ now },
        )
        verifier.verify(v["token"], requested_execution_id: exec_id)
        VectorResult.fail("consent", name, "expected error '#{expect}', got Ok")
      rescue IAIso::Consent::ExpiredTokenError
        return VectorResult.fail("consent", name, "expected '#{expect}', got expired") if expect != "expired"
        VectorResult.pass("consent", name)
      rescue IAIso::Consent::RevokedTokenError
        return VectorResult.fail("consent", name, "expected '#{expect}', got revoked") if expect != "revoked"
        VectorResult.pass("consent", name)
      rescue IAIso::Consent::InvalidTokenError => e
        return VectorResult.fail("consent", name, "expected '#{expect}', got invalid: #{e.message}") if expect != "invalid"
        VectorResult.pass("consent", name)
      rescue => e
        VectorResult.fail("consent", name, "unexpected: #{e.class}: #{e.message}")
      end

      def run_roundtrip(shared_key, v)
        name = "roundtrip/#{v["name"]}"
        issue = v["issue"]
        ttl = (issue["ttl_seconds"] || 3600).to_i
        subject = issue["subject"]
        scopes = issue["scopes"]
        exec_id = issue["execution_id"]
        metadata = issue["metadata"]
        now = (v["now"] || 1_700_000_000).to_i
        issuer = v["issuer"] || "iaiso"
        alg = parse_alg(v)

        is = IAIso::Consent::Issuer.new(
          algorithm: alg, issuer: issuer, hs_key: shared_key,
          clock: ->{ now },
        )
        scope = is.issue(
          subject: subject, scopes: scopes, execution_id: exec_id,
          ttl_seconds: ttl, metadata: metadata,
        )

        expect_success = !!v["expected_after_verify_succeeds"]
        verify_exec = v["verify_with_execution_id"]
        ver = IAIso::Consent::Verifier.new(
          algorithm: alg, issuer: issuer, hs_key: shared_key,
          clock: ->{ now + 1 },
        )
        begin
          verified = ver.verify(scope.token, requested_execution_id: verify_exec)
        rescue => e
          return expect_success ?
            VectorResult.fail("consent", name, "expected verify to succeed, failed: #{e.message}") :
            VectorResult.pass("consent", name)
        end
        return VectorResult.fail("consent", name, "expected verify to fail, succeeded") unless expect_success
        return VectorResult.fail("consent", name, "subject mismatch") if verified.subject != subject
        return VectorResult.fail("consent", name, "scopes mismatch") if verified.scopes != scopes
        VectorResult.pass("consent", name)
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConformance/ConsentRunner.swift
import Foundation
import IAIsoConsent

/// Runs the consent section of the conformance suite.
public enum ConsentRunner {

    public static func run(specRoot: String) throws -> [VectorResult] {
        let path = "\(specRoot)/consent/vectors.json"
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        guard let doc = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return [VectorResult.fail("consent", "load", "could not parse vectors.json")]
        }
        let sharedKeyText = (doc["hs256_key_shared"] as? String) ?? ""
        guard let sharedKey = sharedKeyText.data(using: .utf8) else {
            return [VectorResult.fail("consent", "load", "shared key missing")]
        }

        var out: [VectorResult] = []
        for v in (doc["scope_match"] as? [[String: Any]] ?? []) {
            out.append(runScopeMatch(v))
        }
        for v in (doc["scope_match_errors"] as? [[String: Any]] ?? []) {
            out.append(runScopeMatchError(v))
        }
        for v in (doc["valid_tokens"] as? [[String: Any]] ?? []) {
            out.append(runValidToken(sharedKey: sharedKey, v))
        }
        for v in (doc["invalid_tokens"] as? [[String: Any]] ?? []) {
            out.append(runInvalidToken(sharedKey: sharedKey, v))
        }
        for v in (doc["issue_and_verify_roundtrip"] as? [[String: Any]] ?? []) {
            out.append(runRoundtrip(sharedKey: sharedKey, v))
        }
        return out
    }

    private static func runScopeMatch(_ v: [String: Any]) -> VectorResult {
        let name = "scope_match/\((v["name"] as? String) ?? "?")"
        let granted = (v["granted"] as? [String]) ?? []
        let requested = (v["requested"] as? String) ?? ""
        let want = (v["expected"] as? Bool) ?? false
        do {
            let got = try Scopes.granted(granted, requested)
            return got == want
                ? VectorResult.pass("consent", name)
                : VectorResult.fail("consent", name, "got \(got), want \(want)")
        } catch {
            return VectorResult.fail("consent", name,
                "unexpected exception: \(error)")
        }
    }

    private static func runScopeMatchError(_ v: [String: Any]) -> VectorResult {
        let name = "scope_match_errors/\((v["name"] as? String) ?? "?")"
        let granted = (v["granted"] as? [String]) ?? []
        let requested = (v["requested"] as? String) ?? ""
        let expectErr = (v["expect_error"] as? String) ?? ""
        do {
            _ = try Scopes.granted(granted, requested)
            return VectorResult.fail("consent", name,
                "expected error containing '\(expectErr)', got Ok")
        } catch let err as ScopesError {
            let msg = err.description.lowercased()
            if msg.contains(expectErr.lowercased()) {
                return VectorResult.pass("consent", name)
            }
            return VectorResult.fail("consent", name,
                "expected '\(expectErr)', got: \(err.description)")
        } catch {
            return VectorResult.fail("consent", name,
                "unexpected exception type: \(error)")
        }
    }

    private static func parseAlg(_ v: [String: Any]) -> Algorithm {
        if let s = v["algorithm"] as? String, let a = Algorithm(rawValue: s) {
            return a
        }
        return .hs256
    }

    private static func runValidToken(sharedKey: Data, _ v: [String: Any]) -> VectorResult {
        let name = "valid_tokens/\((v["name"] as? String) ?? "?")"
        let now = Int64((v["now"] as? Int) ?? 0)
        let issuer = (v["issuer"] as? String) ?? "iaiso"
        let alg = parseAlg(v)
        let token = (v["token"] as? String) ?? ""

        let verifier = Verifier(
            algorithm: alg, issuer: issuer, hsKey: sharedKey,
            clock: { now })
        do {
            let scope = try verifier.verify(token)
            guard let exp = v["expected"] as? [String: Any] else {
                return VectorResult.fail("consent", name, "missing expected block")
            }
            if let s = exp["sub"] as? String, scope.subject != s {
                return VectorResult.fail("consent", name, "sub: got \(scope.subject), want \(s)")
            }
            if let j = exp["jti"] as? String, scope.jti != j {
                return VectorResult.fail("consent", name, "jti: got \(scope.jti), want \(j)")
            }
            if let want = exp["scopes"] as? [String] {
                if scope.scopes != want {
                    return VectorResult.fail("consent", name, "scopes mismatch")
                }
            }
            let wantExec = exp["execution_id"] as? String
            if scope.executionId != wantExec {
                return VectorResult.fail("consent", name,
                    "execution_id: got \(scope.executionId ?? "nil"), want \(wantExec ?? "nil")")
            }
            return VectorResult.pass("consent", name)
        } catch {
            return VectorResult.fail("consent", name, "verify failed: \(error)")
        }
    }

    private static func runInvalidToken(sharedKey: Data, _ v: [String: Any]) -> VectorResult {
        let name = "invalid_tokens/\((v["name"] as? String) ?? "?")"
        let now = Int64((v["now"] as? Int) ?? 0)
        let issuer = (v["issuer"] as? String) ?? "iaiso"
        let alg = parseAlg(v)
        let execId = v["execution_id"] as? String
        let expectErr = (v["expect_error"] as? String) ?? ""
        let token = (v["token"] as? String) ?? ""

        let verifier = Verifier(
            algorithm: alg, issuer: issuer, hsKey: sharedKey,
            clock: { now })
        do {
            _ = try verifier.verify(token, requestedExecutionId: execId)
            return VectorResult.fail("consent", name,
                "expected error '\(expectErr)', got Ok")
        } catch let err as ConsentError {
            let actualKind: String
            switch err {
            case .expiredToken: actualKind = "expired"
            case .revokedToken: actualKind = "revoked"
            case .invalidToken: actualKind = "invalid"
            case .insufficientScope: actualKind = "insufficient_scope"
            }
            if actualKind == expectErr {
                return VectorResult.pass("consent", name)
            }
            return VectorResult.fail("consent", name,
                "expected '\(expectErr)', got \(actualKind): \(err.description)")
        } catch {
            return VectorResult.fail("consent", name, "unexpected: \(error)")
        }
    }

    private static func runRoundtrip(sharedKey: Data, _ v: [String: Any]) -> VectorResult {
        let name = "roundtrip/\((v["name"] as? String) ?? "?")"
        guard let issueSpec = v["issue"] as? [String: Any] else {
            return VectorResult.fail("consent", name, "missing issue block")
        }
        let ttl = Int64((issueSpec["ttl_seconds"] as? Int) ?? 3600)
        let subject = (issueSpec["subject"] as? String) ?? ""
        let scopes = (issueSpec["scopes"] as? [String]) ?? []
        let execId = issueSpec["execution_id"] as? String
        let metadata = issueSpec["metadata"] as? [String: Any]
        let now = Int64((v["now"] as? Int) ?? 1_700_000_000)
        let issuer = (v["issuer"] as? String) ?? "iaiso"
        let alg = parseAlg(v)

        let issuerObj = Issuer(
            algorithm: alg, issuer: issuer, hsKey: sharedKey, clock: { now })
        let scope: Scope
        do {
            scope = try issuerObj.issue(
                subject: subject, scopes: scopes,
                executionId: execId, ttlSeconds: ttl, metadata: metadata)
        } catch {
            return VectorResult.fail("consent", name, "issue failed: \(error)")
        }

        let expectSuccess = (v["expected_after_verify_succeeds"] as? Bool) ?? false
        let verifyExec = v["verify_with_execution_id"] as? String

        let verifier = Verifier(
            algorithm: alg, issuer: issuer, hsKey: sharedKey,
            clock: { now + 1 })
        do {
            let verified = try verifier.verify(scope.token, requestedExecutionId: verifyExec)
            if !expectSuccess {
                return VectorResult.fail("consent", name,
                    "expected verify to fail, succeeded")
            }
            if verified.subject != subject {
                return VectorResult.fail("consent", name, "subject mismatch")
            }
            if verified.scopes != scopes {
                return VectorResult.fail("consent", name, "scopes mismatch")
            }
            return VectorResult.pass("consent", name)
        } catch {
            if expectSuccess {
                return VectorResult.fail("consent", name,
                    "expected verify to succeed, failed: \(error)")
            }
            return VectorResult.pass("consent", name)
        }
    }
}
